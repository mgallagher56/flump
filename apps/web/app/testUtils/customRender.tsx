// create a custom render function with ChakraProvider
import { type ReactNode } from 'react';

import { type RenderOptions, type RenderResult, render as rtlRender } from '@testing-library/react';
import userEvent, { type UserEvent } from '@testing-library/user-event';
import { Provider } from '~/components/ui/provider';

const user = userEvent.setup();

const customRender = (ui: ReactNode, options?: RenderOptions): RenderResult & { user: UserEvent } => {
  const wrappedUi = <Provider>{ui}</Provider>;

  return {
    ...rtlRender(wrappedUi, options),
    user
  };
};

export default customRender;
