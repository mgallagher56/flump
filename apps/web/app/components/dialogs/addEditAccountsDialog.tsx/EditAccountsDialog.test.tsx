import type { ReactNode } from 'react';

import { fireEvent } from '@testing-library/react';
import { vi } from 'vitest';
import customRender from '~/testUtils/customRender';

import mockUser from '__mocks__/user';

import AddEditAccountsDialogBtn from './AddEditAccountsDialog';

const mocks = vi.hoisted(() => ({
  mockUseLoaderData: vi.fn(),
  mockUseRevalidator: vi.fn(() => ({ revalidate: vi.fn() })),
  mockFrom: vi.fn(() => ({
    update: () => ({
      eq: () => ({
        eq: () => ({})
      })
    })
  }))
}));

vi.mock('react-router', async () => {
  const actual: Record<string, unknown> = await vi.importActual('react-router');
  return {
    ...actual,
    useLoaderData: mocks.mockUseLoaderData,
    useRevalidator: mocks.mockUseRevalidator,
    Form: ({ children }: { children: ReactNode }) => <form>{children}</form>,
    useSubmit: () => ({ onSubmit: vi.fn() })
  };
});

vi.mock('app/utils/supabase', () => ({
  default: {
    from: mocks.mockFrom
  }
}));

describe('<EditAccountDialogBtn', () => {
  mocks.mockUseLoaderData.mockReturnValue({ user: mockUser });
  const { baseElement, getByText, getAllByText } = customRender(
    <AddEditAccountsDialogBtn isEditAccount accountId="123456" />
  );
  const triggerBtn = getByText('edit');
  expect(triggerBtn).toBeDefined();
  fireEvent.click(triggerBtn);
  test('should render edit account dialog when trigger button is clicked', () => {
    const addAccountBtn = getAllByText('save')[0];
    fireEvent.click(addAccountBtn);
    expect(mocks.mockFrom).toBeCalled();
    const htmlString = baseElement.outerHTML.toString();
    const baseElementConstant = htmlString.replaceAll(/style="[^"]*"/g, '');
    expect(baseElementConstant).toMatchSnapshot();
  });
});
