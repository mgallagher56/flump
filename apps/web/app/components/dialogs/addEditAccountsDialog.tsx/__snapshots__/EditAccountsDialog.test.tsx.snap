// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<EditAccountDialogBtn > should render edit account dialog when trigger button is clicked 1`] = `"<body><div><button type="button" class="chakra-button css-1ltlu47">edit</button></div><div class="chakra-portal"><div class="chakra-modal__overlay css-13d1tkw" ></div><div data-focus-guard="true" tabindex="0" ></div><div data-focus-lock-disabled="false"><div class="chakra-modal__content-container css-1oxhx2p" tabindex="-1"><section class="chakra-modal__content css-6yiygt" role="dialog" id="chakra-modal-:r0:" tabindex="-1" aria-modal="true"  aria-labelledby="chakra-modal--header-:r0:" aria-describedby="chakra-modal--body-:r0:"><header class="chakra-modal__header css-72fd9l" id="chakra-modal--header-:r0:">editAccount</header><button type="button" aria-label="Close" class="chakra-modal__close-btn css-6sdaxy"><svg viewBox="0 0 24 24" focusable="false" class="chakra-icon css-onkibi" aria-hidden="true"><path fill="currentColor" d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"></path></svg></button><div class="chakra-modal__body css-0" id="chakra-modal--body-:r0:"><form><div class="css-1kkt86i"><div class="d_flex flex-d_column gap_2"><label for="Name">Name</label><div><div class="css-3a3hw0"><input placeholder="accountName" id="Name" class="chakra-input css-0" value="" name="name"></div></div></div><div class="d_flex flex-d_column gap_2"><label for="Account Type">Account Type</label><div class="chakra-select__wrapper css-42b2qy"><select name="type" id="Account Type" class="chakra-select css-1ik61og"><option value="Current">Current</option><option value="Saving">Saving</option><option value="Mortgage">Mortgage</option><option value="Loan">Loan</option><option value="Owed">Owed</option></select><div class="chakra-select__icon-wrapper css-162mkon"><svg viewBox="0 0 24 24" role="presentation" class="chakra-select__icon" focusable="false" aria-hidden="true" ><path fill="currentColor" d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"></path></svg></div></div></div></div></form></div><footer class="chakra-modal__footer css-192qrng"><div role="group" class="chakra-button__group css-1ind3v2" data-orientation="horizontal"><button type="button" class="chakra-button css-kn8t1u">close</button><button type="submit" class="chakra-button css-kn8t1u" form="123456">save</button></div></footer></section></div></div><div data-focus-guard="true" tabindex="0" ></div></div></body>"`;
